### DISCUSSION TITLE
Memory Leak

### DISCUSSION THREAD

**Discussion by finaIcutpro** [dotnet/runtime#97723] [General]
Hi, quick question: is there a known memory leak with Microsoft.Speech.Recognition && System.Speech.Recognition?
I seem to be having an issue with a memory leak and have isolated it, profiled it, and confirmed that the memory leak wasn’t caused by my code.

Will provide code if necessary! This is more of a question for if there’s a known issue pertaining to this.

**Comment by finaIcutpro** [comment]
**Description:**
When running the speech recognition using `Microsoft.Speech.Recognition`, there appears to be a memory leak issue. The recognition engine constantly grows in memory during execution, even though it is disposed of correctly. The problem is specifically observed in the following code block:

```csharp
while (!stoppingToken.IsCancellationRequested && !disposed)
{
    Thread.Sleep(100);
    try
    {
        int read = originStream.Read(buffer, 0, 48000);
        bufferedByteStream.Write(buffer, 0, read);
    }
    catch (IOException ex) when (ex.InnerException is SocketException { SocketErrorCode: SocketError.ConnectionReset })
    {
        Console.WriteLine("Connection was forcibly closed by the remote host.");
        bufferedByteStream.Close();
        disposed = true;
    }
}
```

**Context:**
- The recognition engine grows in memory while it runs in a separate thread.
- Disposing of the engine at the end of the execution does not resolve the issue.
- The memory growth is evident even with correct disposal practices.

**Additional Information:**
- The issue happens with`Microsoft.Speech.Recognition` and is also observed in `System.Speech.Recognition`.
- The code is part of a background service and runs continuously.

**Code Snippet for Reference:**
```csharp
// ... (previous code)

using (SpeechRecognitionEngine engine = SREBuilder.Create(new[] { "1FriendlyDoge", "notanoob600m", "RoyalCrests" }))
{
  engine.SpeechRecognized += (sender, e) => HandleSpeechRecognized(sender, e, userId, guildId);
  engine.SetInputToAudioStream(bufferedByteStream, new SpeechAudioFormatInfo(EncodingFormat.Pcm, 48000, 16, 2, 192000, 4, null));
  engine.RecognizeAsync(RecognizeMode.Multiple);
    
    while (!stoppingToken.IsCancellationRequested && !disposed)
    {
        Thread.Sleep(100);
        try
        {
            int read = originStream.Read(buffer, 0, 48000);
            bufferedByteStream.Write(buffer, 0, read);
        }
        catch (IOException ex) when (ex.InnerException is SocketException { SocketErrorCode: SocketError.ConnectionReset })
        {
            Console.WriteLine("Connection was forcibly closed by the remote host.");
            bufferedByteStream.Close();
            disposed = true;
        }
    }
}

```

**Reply by danmoseley** [reply]
I don't know of a leak. Does it leak when you use .NET Framework rather then the Nuget package?

**Reply by finaIcutpro** [reply]
Hi Dan,

I will try this and will get back to you tomorrow, thanks!

**Reply by En3Tho** [reply]
How often exception you expect is thrown? It seems that this bufferedStream can grow quite a lot. How it is used?

Also your snipped doesn't actually show how you use this recognition engine.

If your leak is managed one then you can quite easily track the exact root of a dangling object. Are you sure it is the engine not and this seemingly endlessly (well, until a very specific exception is thrown) growing bufferedStream?

**Reply by finaIcutpro** [reply]
> I don't know of a leak. Does it leak when you use .NET Framework rather then the Nuget package?

@danmoseley I have checked and System.Speech.Recognition is not in the .NET framework.

**Reply by finaIcutpro** [reply]
> How often exception you expect is thrown? It seems that this bufferedStream can grow quite a lot. How it is used?
> 
> Also your snipped doesn't actually show how you use this recognition engine.
> 
> If your leak is managed one then you can quite easily track the exact root of a dangling object. Are you sure it is the engine not and this seemingly endlessly (well, until a very specific exception is thrown) growing bufferedStream?

@En3Tho, the exception gets called when a socket connection closes (each socket connection has its own instance of SpeechRecognitionEngine) and only happens once per instance. After that everything is being disposed.

The buffered steam is just a list of bytes with a predefined size (48,000 bytes in this case, which is 0.25s of audio data) and its size never changes.

When trying to profile the issue using dotMemory the issue seems to be a ton of `Byte[]` object with a stack trace of literally just `[AllThreadsRoot]`.

Thank you both for your contributions so far!

**Reply by finaIcutpro** [reply]
I have edited my original comment to include more relevant code bits. 
ie;
```cs
engine.SpeechRecognized += (sender, e) => HandleSpeechRecognized(sender, e, userId, guildId);
engine.SetInputToAudioStream(bufferedByteStream, new SpeechAudioFormatInfo(EncodingFormat.Pcm, 48000, 16, 2, 192000, 4, null));
engine.RecognizeAsync(RecognizeMode.Multiple);```

**Reply by finaIcutpro** [reply]
Hi,

I am still having a problem with this.
All input is appreciated! 👍 😄

**Reply by En3Tho** [reply]
Honestly it's quite hard to tell what might be the problem here. If you could share a dump or something that would be useful. Your code snippet points to "could be anything" to be honest. Considering you're sure your streams are not growing indefinitely.

It really might be a leak in MSR but it is hard to diagnoze without additional information, ideally a concrete dump. I whink you can share it with an MS employee and be sure data won't leak outside. Well, if you can.

**Reply by finaIcutpro** [reply]
I've moved this issue here: https://github.com/dotnet/runtime/issues/97822
